b2sdk/progress.py

```diff
@@ -51,10 +52,31 @@ def bytes_completed(self, byte_count):
         Transfer can fail and restart from beginning so byte count can
         decrease between calls.
 
-        :param int byte_count: number of bytes have been transferred
+        :param byte_count: number of bytes have been transferred
+        """
+
+    def can_change_description(self) -> bool:
+        """
+        Determines, on a per-implementation basis, whether the description can be changed at this time.
+        """
+        return True
+
+    def change_description(self, new_description: str) -> None:
+        """
+        Ability to change the description after the listener is started.
+
+        Note: whether the change of description is allowed depends on the implementation.
+        The safest option is to change the description before setting the total bytes.
+
+        :param new_description: the new description for the presentation
         """
+        can_change_description = self.can_change_description()
+        assert can_change_description, 'Changing description is not allowed at this point.'
+        # In case assertions are disabled, we will perform the change only when it's allowed.
+        if can_change_description:
+            self.description = new_description
```

Comment:
If we're worried about disabled asserts, why use `assert` here instead of a regular exception? Or, if we want to support simply returning when you can't change description, even a `bool` return type instead of having `can_change_description()` method.

Just genuine questions, not saying this should be changed.

Reply:
The idea about returning the boolean being returned is a fair one.

`assert` was used, as it seems its the only "right" way of checking for sanity in the code around.
`can_change_description()` method is more for internal use than external one.
I like the idea about returning boolean instead of raising assertion! Thank you!

Reply:
I've renamed `can_change_description` to have an underscore in front (protected method) and removed assertion entirely.